pip/pip-338.md

```diff
@@ -0,0 +1,172 @@
+
+Moreover, as per current code provided above, even if the client provides a listenerName/listener header, the redirect urls are taken from the internal listener or the first listener with http protocol which may not be a client-resolvable http address.
+
+# Proposed Solution
+To fix inconsistency with all protocol’s service url, we need to include all broker addresses associated with the listener name while returning the lookupResult, rather than solely the service URL
+
+Also, Introduce a broker-level property to determine the appropriate listener for lookup calls if listenerName is missing from the client request. This can help deal with the side effects of not having listerName on the client side when multiple advertisedListeners are being used.
+
+Also, this can help to determine the broker service URL that needs to be returned in case of lookup redirects and also for using Pulsar admin API calls from outside the network.
+
+This will help in having a more transparent listener identifier for internal and external communications.
+
+
+# Design & Implementation Details
+## Approach
+**1. Fix Code Flow - Inconsistency with listener usage with protocols other than pulsar and pulsar+ssl** 
+
+Currently, lookup result return the broker service url corresponding to the internal listerner in case of http or https protocol urls. To address the issue in the workflow, it is necessary to include all broker addresses associated with the listener name, rather than solely the service URL. There can be a check if all broker addresses corresponding to the given listener is null, then use the default listener(discussed below)
+
+**2. Only return the service URL corresponding to the lookupListener**
+
+This approach introduces one new configuration `lookupListenerName` in the broker.conf. The `lookupListenerName` is used to specify which service URL should be returned to the clients from the lookup requests or redirects. `lookupListenerName` should be present in the `advertisedListeners` list.
+Users can set up the `lookupListenerName`. for example:
+
+```
+existing configs:
+advertisedListeners=internal:pulsar://xyz-broker:6660,internal:pulsar+ssl://192.168.1.11:6651,external:http://192.168.1.11:8080
+internalListenerName=internal
+
+new config required:
+lookupListenerName=external
+```
+
+In the approach, when the client sends a lookup request to the broker to get the owner broker without the listnerName, the broker only returns one service URL that is with the same listener name for which the default value of the lookupListenerName is given in the broker configuration. Therefore, we must allow the client to get the corresponding service URL with the same advertised listener name as present in the broker config.
```

Comment:
This can be covered in the broker code. 
One possibility would be to extend this method to handle redirecting to the URL which the incoming request's active listener name points to: https://github.com/apache/pulsar/blob/3158fd3550f9e3a0b2c0316c92265318b209f4f5/pulsar-broker/src/main/java/org/apache/pulsar/broker/web/PulsarWebResource.java#L1207-L1242

The advertised listeners are available in the BrokerLookupData records which are already used. https://github.com/apache/pulsar/blob/6514cddee5ef1f1ecf8a63fc78429d65cd5ffd36/pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/data/BrokerLookupData.java#L35

Reply:
A method could be added to PulsarWebResource for getting the listener name from a parameter or a header. There's no need to add these explicitly for each method in the API as has been done for the Topic lookup. 

```
    static final String LISTENERNAME_HEADER = "X-Pulsar-ListenerName";
    static final String LISTENERNAME_PARAMETER = "listenerName";
    
    public String listenerName() {
        return Optional.ofNullable(httpRequest.getParameter(LISTENERNAME_PARAMETER))
                .or(() -> Optional.ofNullable(httpRequest.getHeader(LISTENERNAME_HEADER)))
                .orElseGet(() -> pulsar().getConfiguration().getInternalListenerName());
    }
```

Reply:
Primary goal here is also to limit the changes to server side only, because there are cases where there is either no scope of adding a listener name as query param or header in the current system. Even in the systems which can add this config, think 100s of different applications needing either a client upgrade or a config change to add this listener name.

And in case we do end up asking all clients to send a specific listener in their API calls or lookup requests, then isn't that a default for us? Thus, the addition of a default value for the incoming listener name.

Reply:
> because there are cases where there is either no scope of adding a listener name as query param or header in the current system.

In PIP-95 design I think that the idea is that this could be added in a reverse proxy that is in front of Pulsar. 
The alternative is to extend the implementation to support a default listener name based on the bind address/port, as mentioned in the previous comment, https://github.com/apache/pulsar/pull/22039#discussion_r1487413883 .
I'd prefer that approach since it would make the solution consistent with the other parts of PIP-61 & PIP-95.

Reply:
It is becoming more and more clear that at least you are preferring the bind address solution.
The purpose of this PIP was to fix a broken behavior. We have already established that things are broken in the current state, be it a miss/bug or half done implementation.
If this PIP is not the right solve, then we would not like to spend more time on this discussion. We are also not here to pickup and finish someone else's PIP in our current state of urgency with running our production with this bug.

It seems to me that the right approach in our situation is to patch pulsar ourselves and maintain that patch version after version internally.

I wish there was an agreement here but from the start, you have been advocating for a different approach that what our production currently needs.

I will discuss this internally and close this PIP later today. I appreciate your time and effort in the discussion.

Reply:
> It is becoming more and more clear that at least you are preferring the bind address solution.

The bind address solution is part of PIP-61 and PIP-95 design. That is the reason to prefer it. The bind address solution could be extended to cover the http server so that a bind address can be configured to have a specific default listener.

> If this PIP is not the right solve, then we would not like to spend more time on this discussion. We are also not here to pickup and finish someone else's PIP in our current state of urgency with running our production with this bug.

There's a need to keep the project's decisions consistent going forward. Since PIP-61 and PIP-95 already have a specific design which is reasonable, it's better to continue down that path instead of introducing a kludge in the design to do a short cut. If you need to take a short cut, you are free to do so. Apache Pulsar is OSS.

Reply:
> The bind address solution is part of PIP-61 and PIP-95 design. That is the reason to prefer it. The bind address solution could be extended to cover the http server so that a bind address can be configured to have a specific default listener.

Could you please answer this question from the other comment https://github.com/apache/pulsar/pull/22039#discussion_r1487445671

> Could you run us through how that would work for the case when a normal producer client does an http based lookup call only to use the native protocol response from it and to also ensure that redirect in case of 307 response of that lookup call also works correctly?
We have to consider the fact that due to many security/audit reasons, there might only be a single http based load balancer at port 80 possible for the brokers which may only talk to a single broker-port of 8080.

Reply:
> Could you run us through how that would work for the case when a normal producer client does an http based lookup call only to use the native protocol response from it and to also ensure that redirect in case of 307 response of that lookup call also works correctly?

First of all, that would require fixing the bug where the listenerName is ignored, that's mentioned in https://github.com/apache/pulsar/pull/22039#discussion_r1487402439 . The HTTP lookup already supports passing the listener name. That would obviously require passing the listener name from clients. The alternative without making changes to Pulsar would be to put a HTTP reverse proxy such as nginx in front of Pulsar REST calls that sets the "X-Pulsar-ListenerName" header. I believe that adding a header is also possible with nginx ingress controller. It's not optimal, but that's how PIP-95 has been scoped.

Reply:
> That would obviously require passing the listener name from clients. The alternative without making changes to Pulsar would be to put a HTTP reverse proxy such as nginx in front of Pulsar REST calls that sets the "X-Pulsar-ListenerName" header. I believe that adding a header is also possible with nginx ingress controller. It's not optimal, but that's how PIP-95 has been scoped.

My opinion is that a design should not force users to set defaults from outside pulsar systems. Why was it not discussed that this should be done via a config rather than asking either _every_ client to send a value or setup another component or config to add a "required" header at the load balancer layer...

Can we agree that there are cases where things can be miss in old PIP and not all discussed and approved PIPs should be taken as bible :)

Reply:
> My opinion is that a design should not force users to set defaults from outside pulsar systems. Why was it not discussed that this should be done via a config rather than asking either _every_ client to send a value or setup another component or config to add a "required" header at the load balancer layer...

It's part of [PIP-95 design](https://github.com/apache/pulsar/issues/12040) and it is mentioned there.  There are also [comments about the scope limitation](https://github.com/apache/pulsar/pull/12072#issuecomment-921903045).

I agree that it's not currently convenient when there isn't out-of-the-box support without using a reverse proxy to add the header value. That could be addressed with the bindAddresses solution which sets a default for each binding address/port, in the similar way as it has been addressed for the Pulsar binary protocol.

Reply:
just to add to this point that even when listener name is being passed by the clients, as per the current implementation it is not addressed if its corresponding to the http protocol broker's address. The PIP not only focussing on adding a default listener for clients/external communication but also to fix this bug. 
Also when we are saying that internalListenerName is a default listener, to be more precise here- it can be default listener for broker to broker communication not from clients/external communication. That was the reason to add a default listener for external communication as currently we are doing in case of [KIP](https://github.com/apache/pulsar/pull/22039#discussion_r1486401660) as @grssam mentioned.